### PR TITLE
Add Tile Entities

### PR DESCRIPTION
**Pull request by Deamon5550**
This PR adds interfaces for tile entities.

Notably:
- [x] Banner
- [x] Beacon
- [x] BrewingStand
- [x] Chest
- [x] CommandBlock
- [x] Comparator
- [x] DaylightDetector
- [x] Dispenser
- [x] Dropper
- [x] EnchantmentTable
- [x] EnderChest
- [x] EndPortal
- [x] Furnace
- [x] MobSpawner
- [x] Note
- [x] Sign
- [x] Skull

There is also a supertype `Tile`.

Also adds the following supporting types
- Inventory (tile entity supertype)
- Lockable (tile entity supertype)
- BannerPatternShape
- SkullType
- WeightedRandomEntity

I chose not to include interfaces for Pistons (due to their volatile nature), and Flower pots (since they are simply a work around until the world format is updated to support more than 16 metadata values and are fully supported via `BlockState`).

### CODE REVIEW

**Comment by kenzierocks** [comment]
But what about mods?

(still reading, btw)

File: src/main/java/org/spongepowered/api/block/data/Skull.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+import com.google.common.base.Optional;
+
+/**
+ * Represents a skull.
+ */
+public interface Skull {
+
+    /**
+     * Gets the rotation of the skull.
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/block/data/Skull.java]: There needs to be a note about this only applying for skulls placed on top of a block.

Comment (Deamon5550) [inline src/main/java/org/spongepowered/api/block/data/Skull.java]: No even skulls placed against walls still have a rotation, the rotation is just always perpendicular to the wall. Unless you were referring to the setter in which case I might agree with you, not sure the outcome of setting a non-perpendicular rotation on a skull attached to a wall.

Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/block/data/Skull.java]: I meant that the returned value is ambiguous otherwise. For example, does `Direction.NORTH` mean facing north on top of a block, or facing north from the side of a block?

Comment (Deamon5550) [inline src/main/java/org/spongepowered/api/block/data/Skull.java]: Ah you are correct, I hadn't missed that there is a second rotation in the BlockState and that the tile entities rotation is only referenced if the block's rotation is `UP`. I'll fix these docs.

File: src/main/java/org/spongepowered/api/block/data/BrewingStand.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+     */
+    int getRemainingBrewTime();
+
+    /**
+     * Sets the remaining time until the brewing is complete. This will only
+     * have effect if the current items within the brewing stand are valid.
+     * 
+     * @param time The new remaining time, in ticks
+     */
+    void setRemainingBrewTile(int time);
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/block/data/BrewingStand.java]: You have a little typo here Tile -> Time

File: src/main/java/org/spongepowered/api/block/data/Beacon.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+ * Represents a Beacon.
+ */
+public interface Beacon extends Lockable {
+
+    /**
+     * Gets the primary effect provided by this beacon.
+     * 
+     * @return The primary effect
+     */
+    PotionEffectType getPrimaryEffect();
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/block/data/Beacon.java]: For plugin altered Beacons it may be usefull to have the ability to use `List`s as well:
* `getEffects()` 
* `setEffects(List<PotionEffectType>)`

Because there may be more effects allowed in there.

File: src/main/java/org/spongepowered/api/block/data/Beacon.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    /**
+     * Sets the primary effect for this beacon.
+     * 
+     * @param effect The new primary effect
+     */
+    void setPrimaryEffect(PotionEffectType effect);
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/block/data/Beacon.java]: In vanilla minecraft the effects are quite limited to specific types, is this an UI limitation only or is this a forced/fixed limitation? If it is forced/fixed, you should add either an exception, a return value or at least some javadocs that it might not set the `PotionEffectType`. Maybe a method `getAllowedEffects()` would be good in that case as well.

File: src/main/java/org/spongepowered/api/block/data/DaylightDetector.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.spongepowered.api.block.data;
+
+/**
+ * Represents a DaylightDetector.
+ */
+public interface DaylightDetector extends Tile {
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/block/data/DaylightDetector.java]: Why no `setOutputSignal(int)` and `getOutputSignal()` here?

IIRC there is a flag that states whether this `DaylightDetector` is inverted or not, is this a `Block` property or an `Tile` property?

File: src/main/java/org/spongepowered/api/block/data/MobSpawner.java [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+     */
+    void setSpawnRange(short range);
+
+    /**
+     * Sets the next entity type and properties to be spawned.
+     * 
+     * @param type The entity type
+     * @param additionalProperties Additional properties to apply to the entity
+     */
+    void setNextEntityToSpawn(EntityType type, Optional<DataContainer> additionalProperties);
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/block/data/MobSpawner.java]: You should use `@Nullable DataContainer` instead of `Optional<DataContainer>`. This way you don't have to check whether the `additionalProperties` is `null` and you can just use `Optional.fromNullable(additionalProperties)` internally. (And IMHO its easier to use)

File: src/main/java/org/spongepowered/api/block/data/MobSpawner.java [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+     */
+    void setNextEntityToSpawn(WeightedRandomEntity entity);
+
+    /**
+     * Defines a number of {@link WeightedRandomEntity}s from which the type of
+     * each batch will be randomly selected based on the weighting value.
+     * 
+     * @param entities The possible entities
+     */
+    void setPossibleEntitiesToSpawn(WeightedRandomEntity... entities);
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/block/data/MobSpawner.java]: Is there a reason why you used an var arg for the setter, but a `Collection` for the getter?

It would be nice if you add a setter using a `Collection` as well. So that you can grab the value from one spawner and set it to another spawner without converting them.

**Comment by ST-DDT** [comment]
You should add your `SkullType` `BannerShapePattern`... to the GameRegistry as well.

File: src/main/java/org/spongepowered/api/block/meta/SkullType.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.spongepowered.api.block.meta;
+
+/**
+ * Represents a type of skull.
+ */
+public interface SkullType {
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/block/meta/SkullType.java]: A `getName()` method would be nice as well, so that you can display it to the user when used in commands.

File: src/main/java/org/spongepowered/api/block/meta/NotePitch.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+ */
+
+package org.spongepowered.api.block.meta;
+
+import org.spongepowered.api.block.data.Note;
+
+/**
+ * Represents a NotePitch which may be played by a {@link Note} block.
+ */
+public interface NotePitch {
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/block/meta/NotePitch.java]: A `getName()` method would be nice as well, so that you can display it to the user when used in commands.

File: src/main/java/org/spongepowered/api/block/data/Beacon.java (hunk 4 above)
Comment (Deamon5550) [inline src/main/java/org/spongepowered/api/block/data/Beacon.java]: It is a UI limitation only.

File: src/main/java/org/spongepowered/api/block/data/DaylightDetector.java (hunk 5 above)
Comment (Deamon5550) [inline src/main/java/org/spongepowered/api/block/data/DaylightDetector.java]: The power is a property of the block state, the inverted nature is not exposed anywhere, but is instead represented by an entirely different BlockType (see `BlockTypes#DAYLIGHT_DETECTOR_INVERTED`)

File: src/main/java/org/spongepowered/api/block/meta/SkullType.java (hunk 8 above)
Comment (Deamon5550) [inline src/main/java/org/spongepowered/api/block/meta/SkullType.java]: I guess I could add one, I didn't originally as the types have no names in the minecraft source. So it would be purely part of our construct.

File: src/main/java/org/spongepowered/api/block/meta/NotePitch.java (hunk 9 above)
Comment (Deamon5550) [inline src/main/java/org/spongepowered/api/block/meta/NotePitch.java]: same as above, although less sure about this one. Even the names I gave them in `NotePitches` are fairly contrived.

File: src/main/java/org/spongepowered/api/block/data/Beacon.java (hunk 3 above)
Comment (Deamon5550) [inline src/main/java/org/spongepowered/api/block/data/Beacon.java]: I'm fairly sure that this is outside of the scope of changes that we want to make to Beacons. Would require major overwriting of the basic functionality of beacons.

Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/block/data/Beacon.java]: No i don't want to set anything that is not possible yet, if there are too much effects in that list the list will be either truncated or setting it throws an exception. (Of course this would have to be stated in the javadocs)

Comment (Bammerbom) [inline src/main/java/org/spongepowered/api/block/data/Beacon.java]: What if the beacon has no effect? Make it Optional.

**Comment by me4502** [comment]
it may be worth having a method that creates a WeightedRandomEntity based off an Entity and weight, that will copy all the entities data over.

**Comment by bloodmc** [comment]
@Deamon5550 mods will require a custom generic interface for TileEntities that they add to the game.

**Comment by me4502** [comment]
@bloodmc, also need to make sure it supports vanilla TEs as well, in the case that a mod adds further data to one.

**Comment by Deamon5550** [comment]
@bloodmc We have it no? See: [Tile.java](https://github.com/SpongePowered/SpongeAPI/pull/381/files#diff-c911b5e2b5389a109346f63d420bd1dfR1)

**Comment by Zidane** [comment]
Cannot pull this due to conflicts.

File: src/main/java/org/spongepowered/api/block/data/Skull.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+     */
+    void setRotation(Direction rotation);
+
+    /**
+     * If the type of the skull is {@link SkullTypes#PLAYER} then this will
+     * return the associated {@link User}.
+     * 
+     * @return The player
+     */
+    Optional<User> getPlayer();
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/block/data/Skull.java]: This should probably return `GameProfile` because the player of the skull doesn't need to have any persistent data on the server.
